scripts/fetch-configlet_v3

```diff
@@ -32,21 +32,20 @@ else
     HEADER="authorization: Bearer ${GITHUB_TOKEN}"
 fi
 
-FILENAME="canonical_data_syncer-${OS}-${ARCH}.${EXT}"
+SUFFIX="${OS}-${ARCH}.${EXT}"
 
 get_url () {
     curl --header "$HEADER" -s "$LATEST" |
-        awk -v filename=$FILENAME '$1 ~ /browser_download_url/ && $2 ~ filename { print $2 }' |
-        tr -d '"'
+        sed -n "s#\s*\"browser_download_url\": \"\(.*/download/.*/configlet.*${SUFFIX}\)\"\$#\1#p"
```

Comment:
```suggestion
        grep "^\s*\"browser_download_url\": \".*${SUFFIX}\"$" |
        cut -d'"' -f4
```
An alternative that uses `grep` and `cut` instead of `sed`. Note that the `sed` pattern doesn't have to include `configlet` either. But maybe it's a good idea to include it, in case we ever release more than one asset with the same `SUFFIX`.

Some goals here:
- Don't use `jq` - the user might not have it installed.
- Use only portable features of `grep`, `sed`, `awk` etc.

Reply:
> Here's an alternative that uses grep and cut instead of sed. Might be more readable to most people.

It definitely is better than the sed one. What was the issue with the `awk` variant? My sole goal with this bit of code is to make it as stable as possible. I don't think it is likely that github will change its API output, but it also doesn't hurt to be paranoid :) E.g., do we need the `^\s*` prefix for the browser download URL?

>  But maybe it's a good idea to include configlet, in case we ever release more than one asset with the same SUFFIX.

Agreed. This makes sense to me.

Reply:
> What was the issue with the `awk` variant?

The key point is that the old fetch script hardcodes the filename of the asset. See:
https://github.com/exercism/canonical-data-syncer/blob/906ffde760b0c7dd1791ae2632fe39f497c75453/scripts/fetch-canonical_data_syncer#L35-L41

So we shouldn't do exact matching, which wouldn't work when we later rename the assets from e.g.
`configlet_v3-linux-64bit.tgz`
to
`configlet-linux-64bit.tgz`

Therefore this PR currently matches using `configlet.*${SUFFIX}` where SUFFIX is e.g. `linux-64bit.tgz`.

I went with `sed` because:
- It's less powerful than `awk`, and I generally favor the least powerful tool when two tools provide equal correctness. `grep` is even less powerful, but maybe it's more idiomatic to capture than to pipe to `cut`.
- Maybe the average reader/maintainer of this script is better at `sed` than `awk`.

Reply:
Pinging @glennj, @IsaacG, @SaschaMann:

In this PR I'm changing the `fetch-canonical_data_syncer` script to work as the `fetch-configlet_v3` script. It will eventually become the new `fetch-configlet` script that lives in `exercism/configlet` and the track repos so that maintainers can download `configlet`.

I want to grab the value of the `browser_download_url` key from the GitHub API that corresponds to the user's platform. For the configlet repo, see the response here https://api.github.com/repos/exercism/configlet/releases/latest.

The previous version of the script hard-coded the filename, but we don't want to do that here; ideally, this script should stay working after this repo is renamed from `configlet-v3` to `configlet` (after being renamed from `canonical-data-syncer` to `configlet-v3`).

What's the most tasteful, portable command that doesn't require `jq` to be installed?

In a later PR I'll add the changes we recently made to the `fetch-configlet` script in `exercism/github-actions`.

Reply:
> What's the most tasteful, portable command that doesn't require jq to be installed?

If you don't like using grep/sed for it, perhaps include [JSON.sh](https://github.com/dominictarr/JSON.sh)? It's only 200ish lines

Reply:
> If you don't like using grep/sed for it,

I do! I just want to check if there's a better way of doing portable pattern matching with substring extraction. Maybe it's better to do something that involves `case ... esac` or `=~`.

> perhaps include JSON.sh? It's only 200ish lines

Mmm. 200 lines might be short for a JSON parser, but it'd still make this script 4x longer - we want to distribute only one file. And it looks like we'd still need to pipe its output to `grep` etc anyway, at which point we're in a similar position?

```
$ curl -sSf https://api.github.com/repos/exercism/configlet/releases/latest | ./JSON.sh -b | grep browser
["assets",0,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-linux-32bit.tgz"
["assets",1,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-linux-64bit.tgz"
["assets",2,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-linux-arm-v5.tgz"
["assets",3,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-linux-arm-v6.tgz"
["assets",4,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-mac-32bit.tgz"
["assets",5,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-mac-64bit.tgz"
["assets",6,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-windows-32bit.zip"
["assets",7,"browser_download_url"]	"https://github.com/exercism/configlet/releases/download/v3.9.2/configlet-windows-64bit.zip"
```

`shellcheck JSON.sh` also produces quite a few complaints.

Reply:
If we're going to add dependencies, they should be managed as such, not copied in wholesale. It should be added in as a meta-repo.

If you're asking about grep vs sed vs awk, my personal preference would be something slightly more verbose and (for me) more readable: `| awk -F": " -v "suffix=$SUFFIX" '$1 ~ /browser_download_url/ && $2 ~ "releases/download/.*" suffix {gsub(/"/, "", $2); print $2}'`

If you want a pure-bash solution,

```
PATTERN="^ *\"browser_download_url\": \"(.*/releases/download/.*${SUFFIX})\""
curl ... | while read -r ; do [[ $REPLY =~ $PATTERN ]] && echo "${BASH_REMATCH[1]}"; done
```

JSON is becoming ubiquitous these days and `jq` is becoming (or already is) the standard tool for handling JSON on the CLI. How bad would it be to depend on having it?

Reply:
I don't want any dependencies :) 

> JSON is becoming ubiquitous these days and jq is becoming (or already is) the standard tool for handling JSON on the CLI. How bad would it be to depend on having it?

As far as I can tell, there is no guarantee that users will have `jq` installed. We want this script to work out of the box for virtually anyone.

Reply:
If we tell people, especially Windows or Mac users without package managers, to install jq, we might as well tell them to install configlet manually instead

Reply:
I don't want to add a `jq` dependency here. And in the next PR I'll add the headers that request a fixed version of the GitHub API, so we don't have to worry much about robustness.

Thanks for the `awk` and pure-bash solutions. But I'll just go with the `grep` version for now, hoping that it's the most readable for the average maintainer. We can bikeshed it more later if it's ever a problem. I preferred the `sed` command, but the `grep | cut` is fine (as long as we don't release an asset that contains `"` - and we won't).

I kept the `^` in the pattern because it communicates the intention a bit better. And the `[[:space:]]` is marginally more portable than the `\s`.

The main thing is that we avoid features that aren't available on all operating systems, like `grep -P` (see for example [OpenBSD grep](https://man.openbsd.org/grep)).